hpct-ldap-client/src/servers/ldap.py

```diff
@@ -0,0 +1,110 @@
+#!/usr/bin/env python3
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+#
+# servers/ldap.py
+"""Server for client operator."""
+import logging
+
+logger = logging.getLogger(__name__)
+import subprocess
+
+from charms.operator_libs_linux.v0 import apt
+from charms.operator_libs_linux.v1 import systemd
+
+from utils.filedata import FileData
+
+
+class LdapClient:
+    """Server to provide LDAP client charm all functionality needed."""
+
+    packages = ["ldap-utils", "sssd-ldap"]
+    systemd_services = ["sssd"]
+
+    def __init__(self):
+        pass
+
+    def disable(self) -> None:
+        """Disable services."""
+        for name in self.systemd_services:
+            systemd.service_pause(name)
+
+    def enable(self) -> None:
+        """Enable services."""
+        for name in self.systemd_services:
+            systemd.service_resume(name)
+
+    def install(self) -> None:
+        """Install using charmlib apt."""
+        if self.packages:
+            try:
+                apt.update()
+                for name in self.packages:
+                    apt.add_package(name)
+            except:
+                raise Exception(f"failed to install package ({name})")
+
+    def is_enabled(self) -> bool:
+        """Check enabled status of services."""
+        if self.systemd_services:
+            for name in self.systemd_services:
+                if not systemd._systemctl("is-enabled", name, quiet=True):
+                    return False
+
+        return True
+
+    def is_installed(self) -> bool:
+        """Check packages are installed."""
+        if self.packages:
+            for name in self.packages:
+                if not apt.DebianPackage.from_installed_package(name).present:
+                    return False
+
+        return True
+
+    def is_running(self) -> bool:
+        """Check running/active status of services."""
+        if self.systemd_services:
+            for name in self.systemd_services:
+                if not systemd.service_running(name):
+                    return False
+
+        return True
+
+    def restart(self) -> None:
+        """Restart servers/services."""
+        self.stop()
+        self.start()
+
+    def start(self) -> None:
+        """Start services."""
+        for name in self.systemd_services:
+            systemd.service_start(name)
+
+    def stop(self) -> None:
+        """Stop services."""
+        for name in self.systemd_services:
+            systemd.service_stop(name)
+
+    def tls_save(self, ca_cert, sssd_conf):
+        """Save tls certificate and sssd conf.
+
+        Parameters
+        ----------
+        ca_cert : str
+            CA certificate.
+        sssd_conf : str
+            SSSD configuration file.
+        """
+        fd = FileData(ca_cert)
+        fd.save("/usr/local/share/ca-certificates/mycacert.crt")
```

Comment:
From https://ubuntu.com/server/docs/service-sssd-ldap 

"If using a custom CA, an easy way to have a host trust it is to place it in /usr/local/share/ca-certificates/ with a .crt extension and run sudo update-ca-certificates."

So for the charm the server creates it's own CA so I think it falls under the custom route. If the charm were to use a public CA then it would instead need to edit `/etc/ldap/ldap.conf` and point `TLS_CACERT` to the CA public file.